File: library/src/main/java/com/devbrackets/android/androidmarkup/parser/markdown/MarkdownElement.java
package com.devbrackets.android.androidmarkup.parser.markdown;

import android.support.annotation.Nullable;

/**
 *
 */
public class MarkdownElement {

    @Nullable
    protected MarkdownElement parent;
    protected MarkdownElement[] children;
    protected int spanType; //see SpanType.java? probably a different parsing type
    protected String text;

    public MarkdownElement(@Nullable MarkdownElement parent, int spanType, String text) {
        this.parent = parent;
        this.spanType = spanType;
        this.text = text;
    }
}

File: library/src/main/java/com/devbrackets/android/androidmarkup/parser/markdown/MarkdownElement.kt
package com.devbrackets.android.androidmarkup.parser.markdown

/**
 * `text` is always null except when the element has no children.  If an element  has children
 * the un-styled text should be a child to simplify ordering
 */
class MarkdownElement(var parent: MarkdownElement?, var spanType: Int, var text: String?) {
    lateinit var children: MutableList<MarkdownElement>

    init {
        children = mutableListOf();
    }

    constructor(parent: MarkdownElement?, spanType: Int, text: String?, children : MutableList<MarkdownElement>) : this(parent, spanType, text) {
        this.children = children
    }

    fun addChild(child: MarkdownElement) {
        child.parent = this
        children.add(child)
    }
}
